src/main/flight/pid.c

```diff
@@ -436,28 +439,94 @@ STATIC_UNIT_TESTED FAST_CODE_NOINLINE float calcHorizonLevelStrength(void)
 // The impact is possibly slightly slower performance on F7/H7 but they have more than enough
 // processing power that it should be a non-issue.
 STATIC_UNIT_TESTED FAST_CODE_NOINLINE float pidLevel(int axis, const pidProfile_t *pidProfile, const rollAndPitchTrims_t *angleTrim,
-                                                        float currentPidSetpoint, float horizonLevelStrength)
+                                                        float rawSetpoint, float horizonLevelStrength, bool newRcFrame)
 {
     const float levelAngleLimit = pidProfile->levelAngleLimit;
     // calculate error angle and limit the angle to the max inclination
     // rcDeflection is in range [-1.0, 1.0]
-    float angle = levelAngleLimit * getLevelModeRcDeflection(axis);
+    float angleTarget = levelAngleLimit * getLevelModeRcDeflection(axis);
 #ifdef USE_GPS_RESCUE
-    angle += gpsRescueAngle[axis] / 100; // ANGLE IS IN CENTIDEGREES
+    angleTarget += gpsRescueAngle[axis] / 100; // ANGLE IS IN CENTIDEGREES
+#endif
+    const float currentAngle = (attitude.raw[axis] - angleTrim->raw[axis]) / 10.0f;
+    const float angleError = angleTarget - currentAngle;
+    const float angleErrorCorrection = angleError * pidRuntime.levelGain;
+    float currentAngleDerivative = pidRuntime.angleDerivative[axis];
+    if(newAttitudeData){
+        currentAngleDerivative = (currentAngle - pidRuntime.previousAngle[axis]) * acc.sampleRateHz / (float)imuConfig()->imu_process_denom; //TODO: parametrise attitude update frequency
+        currentAngleDerivative = pt3FilterApply(&pidRuntime.angleDerivativePt3[axis], currentAngleDerivative);
```

Comment:
This PT3 would not be needed if the `currentAngle` value was smoothed with the `attitudeFilter` as suggested above.

Reply:
I am not sure this is true as current angle does not update at PID loop rate. Like with feedforward and RC rate we need to handle updates to attitude at half the accelerometer sample rate (~500Hz).